src/sfizz/LFO.cpp

```diff
@@ -0,0 +1,295 @@
+    impl.sampleHoldMem_[nth] = sampleHoldValue;
+}
+
+void LFO::processSteps(absl::Span<float> out)
+{
+    unsigned nth = 0;
+    Impl& impl = *impl_;
+    const LFODescription& desc = *impl.desc_;
+    const LFODescription::Sub& sub = desc.sub[nth];
+    const size_t numFrames = out.size();
+
+    const LFODescription::StepSequence& seq = *desc.seq;
+    const float* steps = seq.steps.data();
+    unsigned numSteps = seq.steps.size();
+
+    if (numSteps <= 0)
+        return;
+
+    float samplePeriod = 1.0f / impl.sampleRate_;
+    float phase = impl.subPhases_[nth];
+    float baseFreq = desc.freq;
+    float offset = sub.offset;
+    float ratio = sub.ratio;
+    float scale = sub.scale;
+
+    for (size_t i = 0; i < numFrames; ++i) {
+        float step = steps[static_cast<int>(phase * numSteps)];
+        out[i] += offset + scale * step;
+
+        // TODO(jpc) lfoN_count: number of repetitions
+
+        float incrPhase = ratio * samplePeriod * baseFreq;
+        phase += incrPhase;
+        int numWraps = (int)phase;
+        phase -= numWraps;
+    }
+
+    impl.subPhases_[nth] = phase;
+}
+
+void LFO::process(absl::Span<float> out)
+{
+    Impl& impl = *impl_;
+    const LFODescription& desc = *impl.desc_;
+    size_t numFrames = out.size();
+
+    fill(out, 0.0f);
+
+    size_t skipFrames = std::min(numFrames, impl.delayFramesLeft_);
+    if (skipFrames > 0) {
+        impl.delayFramesLeft_ -= skipFrames;
+        out.remove_prefix(skipFrames);
+        numFrames -= skipFrames;
+    }
+
+    unsigned subno = 0;
+    const unsigned countSubs = desc.sub.size();
+
+    if (countSubs < 1)
+        return;
+
+    if (desc.seq) {
+        processSteps(out);
+        ++subno;
+    }
+
+    for (; subno < countSubs; ++subno) {
+        switch (desc.sub[subno].wave) {
+        case LFOWave::Triangle:
+            processWave<LFOWave::Triangle>(subno, out);
+            break;
+        case LFOWave::Sine:
+            processWave<LFOWave::Sine>(subno, out);
+            break;
+        case LFOWave::Pulse75:
+            processWave<LFOWave::Pulse75>(subno, out);
+            break;
+        case LFOWave::Square:
+            processWave<LFOWave::Square>(subno, out);
+            break;
+        case LFOWave::Pulse25:
+            processWave<LFOWave::Pulse25>(subno, out);
+            break;
+        case LFOWave::Pulse12_5:
+            processWave<LFOWave::Pulse12_5>(subno, out);
+            break;
+        case LFOWave::Ramp:
+            processWave<LFOWave::Ramp>(subno, out);
+            break;
+        case LFOWave::Saw:
+            processWave<LFOWave::Saw>(subno, out);
+            break;
+        case LFOWave::RandomSH:
+            processSH<LFOWave::RandomSH>(subno, out);
+            break;
+        }
+    }
+
+    float fadeIn = impl.fadeInMemory_;
+    const float fadeInPole = impl.fadeInPole_;
+    for (size_t i = 0; i < numFrames; ++i) {
+        out[i] *= fadeIn;
+        fadeIn = fadeInPole * fadeIn + (1 - fadeInPole);
```

Comment:
This is the fade in used in ARIA/Cakewalk?

Reply:
Fade is too slow, I'll adjust the ratio to make this better.

Sfizz-Cakewalk and Sfizz-Aria comparison
![sfizz-cakewalk](https://user-images.githubusercontent.com/17614485/89967668-5cc40280-dc52-11ea-9e8a-55bcccd4a44b.png)
![sfizz-aria](https://user-images.githubusercontent.com/17614485/89967688-66e60100-dc52-11ea-9d0c-b65b3c3ad9c1.png)

Reply:
As it turns out, the fade-in is just a linear progression. Implemented in latest.
(as a note, it would be trivial to let fade-in be defined by curve, if that might be desired)